src/widget/textbox.rs

```diff
@@ -90,30 +199,63 @@ impl Widget<String> for TextBox {
         );
 
         paint_ctx.fill(clip_rect, &BACKGROUND_GREY_LIGHT);
-        paint_ctx.stroke(clip_rect, &border_color, BORDER_WIDTH);
-
-        // Paint the text
-        let text = paint_ctx.text();
-        let text_layout = self.get_layout(text, FONT_SIZE, data);
 
-        let text_height = FONT_SIZE * 0.8;
-        let text_pos = Point::new(0.0 + PADDING_LEFT, text_height + PADDING_TOP);
-
-        // Render text and cursor inside a clip
+        // Render text, selection, and cursor inside a clip
         paint_ctx
             .with_save(|rc| {
                 rc.clip(clip_rect);
 
+                // Layout and measure text
+                let text = rc.text();
+                let text_layout = self.get_layout(text, FONT_SIZE, data);
+
+                let text_height = FONT_SIZE * 0.8;
+                let text_pos = Point::new(0.0 + PADDING_LEFT, text_height + PADDING_TOP);
+
+                let cursor_x = self.substring_measurement_hack(rc, data, 0, self.cursor());
+
                 // If overflowing, shift the text
-                if text_layout.width() + (PADDING_LEFT * 2.) > self.width {
-                    let offset = text_layout.width() - self.width + (PADDING_LEFT * 2.) + 1.;
-                    rc.transform(Affine::translate(Vec2::new(-offset, 0.)));
+                let padding = PADDING_LEFT * 2.;
+                if cursor_x > self.width + self.hscroll_offset - padding {
```

Comment:
In *general* I like paint to be a more or less pure function that just paints the view; this logic feels like it should be split out and invoked from anything that causes a change (event, update). But that's more of a feeling; paint is in fact allowed to do stuff like this.

Reply:
That's a good rule of thumb. I moved it to a method and am calling from event now. In doing so, I realized that there were definitely times I was doing this calculation that I didn't need to be doing it!